Added Zone Wand to make setting Zones easy!

Lunar0sh opened the pull request:
I tested it, it worked also added rules so it always Works!

Then I went insane added a Wiki.

Any questions?
Just ask please

Lunar0sh commented:
Okay tested it again with the width of the Train... what *can* be really buggy... But the Height and Length works perfect!

ekulxam commented on src/main/java/dev/doctor4t/trainmurdermystery/command/ForceResetTrainCommand.java:
```diff
@@ -0,0 +1,33 @@
+
+                    if (retryNeeded) {
+                        // With the fix, this now means "Load requested"
+                        source.sendFeedback(() -> Text.literal("Chunks were unloaded. Loading them now... run this command again in a second!"), false);
```
This should be translatable.

ekulxam commented on src/main/java/dev/doctor4t/trainmurdermystery/game/GameFunctions.java:
```diff
@@ -473,15 +490,24 @@ public static boolean tryResetTrain(ServerWorld serverWorld) {
                 body.discard();
             }
-            for (ItemEntity item : serverWorld.getEntitiesByType(EntityType.ITEM, playerBodyEntity -> true)) {
+            for (net.minecraft.entity.ItemEntity item : serverWorld.getEntitiesByType(net.minecraft.entity.EntityType.ITEM, playerBodyEntity -> true)) {
```
Fix your imports

ekulxam replied:
Again, fix your imports

ekulxam replied:
You should be importing net.minecraft.entity.ItemEntity in your imports

Lunar0sh commented:
Okay, addepted to the Changes you said me to make! 

Without the Train_Center lock, it may or may not tends to get a lil buggy!!

`Also I hope I did the PR right! (Cause i didn't Resolve the conversation)`

Lunar0sh commented on src/main/java/dev/doctor4t/trainmurdermystery/command/SetZoneCommand.java:
```diff
@@ -45,51 +41,31 @@ private static int setZone(CommandContext<ServerCommandSource> context, ZoneType
 
         if (posA == null || posB == null) {
-            source.sendError(Text.literal("Positions not set! Use Right-Click (A) and Shift-Right-Click (B) with the wand."));
+            source.sendError(Text.translatable("commands.trainmurdermystery.setzone.error.no_positions"));
```
I personally tink that the "Right-Click and Shift-Right-Click" is easier to understand.

Okay now I see what you mean, yeah xD
*Just edited it*

Lunar0sh commented:
With the really buggy, i mean that the placement for the Playtrain can be "really" off. 
Im just gonna attach an Image, to make it easier to explain.

<img width="1951" height="1193" alt="image" src="https://github.com/user-attachments/assets/050e4b87-6ef8-443d-9289-72e110e55bdd" />

ekulxam commented:
> With the really buggy, i mean that the placement for the Playtrain can be "really" off. Im just gonna attach an Image, to make it easier to explain.
> 
> <img alt="image" width="1951" height="1193" src="https://private-user-images.githubusercontent.com/121179687/519628569-050e4b87-6ef8-443d-9289-72e110e55bdd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQzODAwNTAsIm5iZiI6MTc2NDM3OTc1MCwicGF0aCI6Ii8xMjExNzk2ODcvNTE5NjI4NTY5LTA1MGU0Yjg3LTZlZjgtNDQzZC05Mjg5LTcyZTExMGU1NWJkZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTEyOVQwMTI5MTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNDBmNjMxNGMzN2E5ZTY4OWJiNmVlOWYwNDMyNGZlYzI2MDAzYWU1ZTNlYjcyYjYzZWRkYjhjZmM0MzIzMTE5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.K9OyQjJSxkmpagPf5qskEEdKcHNHIa4W_yHv5nD_Omg">
> But I think setting it right is more like Try and Error. Also the Main Purpose is to set the Height and Length of the Train.

Is this fixable? I'm not quite sure why this behaviour exists.

ekulxam commented on HowToUseZoneWand.md:
```diff
@@ -0,0 +1,66 @@
+* **Note:** If the chunks are unloaded (too far away), the command will tell you it is loading them. simply wait 1 second and run the command again.
+
+### Common Issues
+* **"Train is not symmetric!"**: Check your Z coordinates. Calculate the distance from -536 for both A and B positions.
```
Would this break trains that have an odd width?

ekulxam commented on src/main/generated/assets/trainmurdermystery/lang/en_us.json:
```diff
@@ -308,6 +308,7 @@
   "item.trainmurdermystery.scorpion": "Scorpion",
   "item.trainmurdermystery.scorpion.tooltip": "Slip in a bed to poison the next person looking for a rest",
   "item.trainmurdermystery.thrown_grenade": "Thrown Grenade",
+  "item.trainmurdermystery.zone_wand": "Zone Wand",
```
Add these to the language provider, otherwise your changes will be overwritten.

Lunar0sh replied:
Okay added to TMMLangGen

ekulxam commented on src/main/java/dev/doctor4t/trainmurdermystery/command/SetZoneCommand.java:
```diff
@@ -0,0 +1,85 @@
+        dispatcher.register(CommandManager.literal("tmm:setzone")
+                .requires(source -> source.hasPermissionLevel(2))
+
+                .then(CommandManager.literal("playArea")
```
Would an EnumArgumentType be better here?

Lunar0sh replied:
Added Enum, but I hadn't time to test it.

ekulxam commented on src/main/java/dev/doctor4t/trainmurdermystery/game/GameFunctions.java:
```diff
@@ -255,6 +257,20 @@ public static void finalizeGame(ServerWorld world) {
+        for (int x = minX; x <= maxX; x++) {
+            for (int z = minZ; z <= maxZ; z++) {
+                // PORTAL ticket keeps chunk loaded for a short time (good for temporary ops)
+                world.getChunkManager().addTicket(ChunkTicketType.PORTAL, new ChunkPos(x, z), 2, new BlockPos(x << 4, 0, z << 4));
```
I'm not sure about using `PORTAL` for this. Perhaps we could create a custom `ChunkTicketType` instead? I'd like to hear your reasoning.

Does `PORTAL` remove itself after a certain duration?

ekulxam commented on src/main/java/dev/doctor4t/trainmurdermystery/game/GameFunctions.java:
```diff
@@ -473,15 +490,24 @@ public static boolean tryResetTrain(ServerWorld serverWorld) {
+                forceLoadBox(serverWorld, trainBox);
+
                 return true;
+                // --- NEW LOGIC ENDS HERE ---
```
Nit: remove

Lunar0sh replied:
Removed with latest commit.

ekulxam commented on src/main/java/dev/doctor4t/trainmurdermystery/index/TMMItems.java:
```diff
@@ -42,6 +42,8 @@
     Item CHAMPAGNE = registrar.create("champagne", new CocktailItem(new Item.Settings().maxCount(1).food(FoodComponents.HONEY_BOTTLE)), EQUIPMENT_GROUP);
     Item NOTE = registrar.create("note", new NoteItem(new Item.Settings().maxCount(4)), EQUIPMENT_GROUP);
 
+    Item ZONE_WAND = registrar.create("zone_wand", new ZoneWandItem(new Item.Settings().maxCount(1)), EQUIPMENT_GROUP);
```
Do we need to give this glint to distinguish it from the vanilla item?

Lunar0sh replied:
I was thinking about giving it a Glint too, but as it's just a Building Item what is Creative only, I don't really think so.

Lunar0sh replied:
Okay, added a Glint anyways...

Lunar0sh commented on src/main/java/dev/doctor4t/trainmurdermystery/game/GameFunctions.java:
```diff
@@ -473,15 +490,24 @@ public static boolean tryResetTrain(ServerWorld serverWorld) {
                 body.discard();
             }
-            for (ItemEntity item : serverWorld.getEntitiesByType(EntityType.ITEM, playerBodyEntity -> true)) {
+            for (net.minecraft.entity.ItemEntity item : serverWorld.getEntitiesByType(net.minecraft.entity.EntityType.ITEM, playerBodyEntity -> true)) {
```
Huh?
What do you mean?
I changed it from `ItemEntity item,` to `net.minecraft.entity.ItemEntity item`

Lunar0sh commented on src/main/java/dev/doctor4t/trainmurdermystery/game/GameFunctions.java:
```diff
@@ -473,15 +490,24 @@ public static boolean tryResetTrain(ServerWorld serverWorld) {
                 }
             } else {
-                TMM.LOGGER.info("Train reset failed: Clone positions not loaded. Queueing another attempt.");
+                // --- NEW LOGIC STARTS HERE ---
```
Removed with latest commit.

Lunar0sh commented:
Okay, so for now, I think I fixed and changed everything what you mentioned... but the Imports, there already is 
`import net.minecraft.entity.ItemEntity;`.
